Daemon cli

cryptoAtwill opened the pull request:
This PR adds the `daemon` CLI to launch the ipc agent json rpc server. It takes the config file path as command line input. Sample config file is placed in `examples` folder.

# Launch
```
cargo build --release
ipc-client daemon ./examples/sample_config.toml
```
@adlrocha While this PR is being reviewed, I will create another one that handles the versioning of the command line from git commit, to keep the PR small.

adlrocha commented on src/cli/commands/mod.rs:
```diff
@@ -0,0 +1,75 @@
+
+/// The `cli` method exposed to handle all the cli commands, ideally from main.
+///
+/// Consider using macro_rules! to handle the registration of new CLI handlers.
```
This is no longer the case, right? We don't use macros.

cryptoAtwill replied:
removed, resolving this

adlrocha commented on src/cli/commands/server.rs:
```diff
@@ -0,0 +1,38 @@
+        log::debug!("launching json rpc server with args: {:?}", arguments);
+
+        let config = Config::from_file(&arguments.config_file)?.server;
+        log::info!("starting json rpc server at: {:}", config.json_rpc_address);
```
```suggestion
        log::info!("starting IPC-agent daemon at: {:}", config.json_rpc_address);
```

adlrocha commented on src/cli/commands/server.rs:
```diff
@@ -0,0 +1,38 @@
+pub(crate) struct LaunchJsonRPC;
+
+#[async_trait]
+impl CommandLineHandler for LaunchJsonRPC {
```
Let's call this `LaunchDaemon` or something more generic, in the future we may launch with this CLI more processes, not only the jsonrpc server
```suggestion
impl CommandLineHandler for LaunchDaemon {
```

cryptoAtwill replied:
updated, resolving this

adlrocha commented on src/cli/commands/server.rs:
```diff
@@ -0,0 +1,38 @@
+use crate::server::jsonrpc::JsonRPCServer;
+
+/// The command to start the ipc agent json rpc server in the foreground.
+pub(crate) struct LaunchJsonRPC;
```
```suggestion
pub(crate) struct LaunchDaemon;
```

cryptoAtwill replied:
updated, resolving this

adlrocha commented on src/cli/commands/server.rs:
```diff
@@ -0,0 +1,38 @@
+//! The Daemon command line handler that prints the info about IPC Agent.
```
Call this file daemon.rs better

cryptoAtwill replied:
updated

adlrocha commented on src/cli/commands/mod.rs:
```diff
@@ -0,0 +1,75 @@
+/// to the current mode. Register a new command accordingly.
+#[derive(Debug, Subcommand)]
+enum Commands {
+    /// Launch the ipc agent json rpc server.
```
```suggestion
    /// Launch the ipc agent daemon.
```

adlrocha commented on src/cli/commands/server.rs:
```diff
@@ -0,0 +1,38 @@
+}
+
+#[derive(Debug, Args)]
+#[command(about = "Prints info of the the ipc agent")]
```
This comment is wrong, right? This should be fixed.

cryptoAtwill replied:
udpated

adlrocha commented on examples/sample_config.toml:
```diff
@@ -0,0 +1,16 @@
+[server]
```
Instead of making this a sample_config, what about making it a `template.toml` in the `config/` for now. I am generally not a fan of this kind of sample files :)

cryptoAtwill replied:
updated

adlrocha commented on src/cli/commands/mod.rs:
```diff
@@ -0,0 +1,75 @@
+    let args = IPCAgentCliCommands::parse();
+
+    let r = match &args.command {
+        Commands::Daemon(args) => <server::LaunchJsonRPC as CommandLineHandler>::handle(args).await,
```
I think I asked you at some point, but is there an alternative to make this a bit less verbose?

cryptoAtwill replied:
updated, simplified the notations.

adlrocha reviewed (changes requested):
First quick round of suggestions. Let me know your thoughts. Also, I see in the description that you were going to handle the version issue in a follow-up PR, let me know if this is still the case (it is not on the critical path so feel free to track it on an issue and focus on the subnet manager for now).

cryptoAtwill commented:
@adlrocha I have updated the code based on the review comments, feel free to take a look. As for the injection of version, I have added the issue: #26 to track it.

adlrocha reviewed (approved):
LGTM. Thanks! I added one last nit but it shouldn't be a blocker. Let's see if @hmoniz wants to have a quick look before merging. We should at some point think about how we are going to add tests to cli commands.

cryptoAtwill commented on src/cli/commands/mod.rs:
```diff
@@ -0,0 +1,74 @@
+struct IPCAgentCliCommands {
+    #[command(subcommand)]
+    command: Commands,
+}
```
udpated and resolving this.
